### PR TITLE
feat(forms): Add cron task to purge old form drafts

### PR DESCRIPTION
**Pull request by ccailly**
| Q | A
|------------- | --- 
| Bug fix?      | no 
| New feature?  | yes 
| BC breaks?    | no 
| Deprecations? | no 
| Tests pass?   | yes 
| Fixed tickets |

When a user creates a form, the object is directly created in base with the `draft` flag.
If the user doesn't save the form, the object will remain in base but will no longer be accessible.
This task deletes old draft forms.

### CODE REVIEW

File: src/Form/Form.php [hunk 1]
```diff
@@ -245,6 +245,37 @@ public function cleanDBonPurge()
...
+     *
+     * @return int
+     * @used-by CronTask
+     */
+    public static function cronPurgeDraftForms($task = null)
+    {
+        $form = new Form();
+        $form->deleteByCriteria([
+            'is_draft' => 1,
+            'date_mod' => ['<', date('Y-m-d H:i:s', strtotime('-1 day'))],
```
Comment (AdrienClairembault) [inline src/Form/Form.php:273]: I would prefer a few more days to be extra safe (original code was 7 days).

```suggestion
            'date_mod' => ['<', date('Y-m-d H:i:s', strtotime('-7 day'))],
```

**Review by AdrienClairembault — DISMISSED** [review]
You'll need to delete the old code in `front/from.form.from.php`:

image

File: install/empty_data.php [hunk 2]
```diff
@@ -921,7 +921,19 @@ public function getEmptyData(): array
                 'logs_lifetime' => 30,
                 'hourmin' => 0,
                 'hourmax' => 6,
-            ],
+            ], [
+                'id' => 48,
+                'itemtype' => 'Glpi\Form\Form',
+                'name' => 'purgedraftforms',
+                'frequency' => DAY_TIMESTAMP,
+                'param' => 30,
```
Comment (cedric-anne) [inline install/empty_data.php:929]:
```suggestion
                'param' => 7,
```

File: install/migrations/update_10.0.x_to_11.0.0/form.php [hunk 3]
```diff
@@ -215,3 +215,12 @@
...
     $migration->addField("glpi_forms_destinations_formdestinations", "config", "JSON NOT NULL COMMENT 'Extra configuration field(s) depending on the destination type'");
 }
+
+CronTask::register('Glpi\Form\Form', 'purgedraftforms', DAY_TIMESTAMP, [
+    'state'         => CronTask::STATE_WAITING,
+    'mode'          => CronTask::MODE_INTERNAL,
+    'hourmin'       => 0,
+    'hourmax'       => 24,
+    'logs_lifetime' => 30,
+    'param'         => 30
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_11.0.0/form.php:225]:
```suggestion
    'param'         => 7
```

File: src/Form/Form.php [hunk 4]
```diff
@@ -245,6 +245,37 @@ public function cleanDBonPurge()
...
+
+    /**
+     * Cron action to purge old form drafts
+     *
+     * @param CronTask|null $task
+     *
+     * @return int
+     * @used-by CronTask
+     */
+    public static function cronPurgeDraftForms($task = null)
```
Comment (cedric-anne) [inline src/Form/Form.php:268]:
```suggestion
    public static function cronPurgeDraftForms(CronTask $task): int
```

File: src/Form/Form.php [hunk 5]
```diff
@@ -245,6 +245,37 @@ public function cleanDBonPurge()
...
+    {
+        $form = new Form();
+        $form->deleteByCriteria([
+            'is_draft' => 1,
+            'date_mod' => ['<', date('Y-m-d H:i:s', strtotime('-7 day'))],
+        ], true);
```
Comment (cedric-anne) [inline src/Form/Form.php:274]:
```suggestion
        $delay = (int) $task->fields['param'];
        $form = new Form();
        $form->deleteByCriteria([
            'is_draft' => 1,
            'date_mod' => ['<', date('Y-m-d H:i:s', strtotime(sprintf('-%d day', $delay)))],
        ], true);
```

File: src/Form/Form.php [hunk 6]
```diff
@@ -245,6 +245,37 @@ public function cleanDBonPurge()
...
+    /**
+     * Give cron information
+     *
+     * @param string $name  Task's name
+     *
+     * @return array Array of information
+     **/
+    public static function cronInfo($name)
+    {
+        return ['description' => __('Purge old form drafts')];
```
Comment (cedric-anne) [inline src/Form/Form.php:257]:
```suggestion
        return [
            'description' => __('Purge old form drafts'),
            'parameter'   => __('Form drafts retention period (in days)'),
        ];
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]
The cron `param` could be used to adjust the retention period of the drafts.

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**